Image GIF pausing

justinmc opened the pull request:
Flutter needs a strong story around pausing animated GIFs because it affects accessibility via [MediaQueryData.disableAnimations](https://main-api.flutter.dev/flutter/widgets/MediaQueryData/disableAnimations.html). See https://github.com/flutter/flutter/issues/175516.

### Today

Currently, it's possible to pause a gif using TickerProvider. However, if paused on initial load, the first frame will never display, and I believe that's a bug.

```dart
TickerMode(
  // Pauses the gif, but if set to false on first load, the first frame never loads.
  enabled: false,
  child: Image(image: myAnimatedGif),
),
```

Currently `disableAnimations` has no effect:

```dart
MediaQuery(
  // Currently has no effect on the Image, until this PR.
  data: MediaQueryData(disableAnimations: true),
  child: Image(image: myAnimatedGif),
),
```

### With this PR

  * MediaQueryData.disableAnimations can be used to pause a gif in the same way that TickerMode can be used today.
  * If either MediaQuery.disableAnimations or TickerMode cause a gif to be paused on first load, it will show its first frame before pausing. This is potentially a breaking change.
  * The docs clearly explain this behavior and how to pause a gif.

Fixes https://github.com/flutter/flutter/issues/175516
Partial fix for https://github.com/flutter/flutter/issues/130976
Fixes Google b/419605327

justinmc committed:
Explain how to use TickerMode to pause animated gifs in the docs for Image

justinmc committed:
Pause for MediaQuery.disableAnimations, too

justinmc committed:
Same tests as TickerMode

justinmc committed:
Loads first frame (TODO, test)

justinmc committed:
Test frame-by-frame for disableAnimations, and initial frame for it and TickerMode

justinmc committed:
Docs

justinmc commented:
@evantea Let me know if this won't solve your problem!

justinmc committed:
Docs typo

justinmc committed:
Fix bug where first 2 frames were rendered

justinmc committed:
Spelling

chunhtai commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -1100,6 +1119,15 @@ class _ImageState extends State<Image> with WidgetsBindingObserver {
+  bool _isPaused = false;
+
+  /// False when the class first is instantiated and true forever after the
+  /// first frame of the image is received.
```
What if image source changes?

justinmc replied:
I think you're right, if the image source changes we  want to load the first frame of the new image. I'll do that.

justinmc committed:
Ignore leaks on tests that use stream completer

justinmc commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -1123,7 +1151,9 @@ class _ImageState extends State<Image> with WidgetsBindingObserver {
     _resolveImage();
 
-    if (TickerMode.of(context)) {
+    _isPaused = !TickerMode.of(context) || (MediaQuery.maybeDisableAnimationsOf(context) ?? false);
```
I've been thinking about them as two different things: MediaQueryData.disableAnimations is related to the system accessibility setting while TickerMode is about definitively disabling all animations. @evantea was unable to use TickerMode because it also disabled the fade animation that displayed the image, but that is not a problem for accessibility (see https://github.com/flutter/flutter/issues/175516#issuecomment-3363396318).

chunhtai replied:
> MediaQueryData.disableAnimations is related to the system accessibility setting

Yes, but what if a user toggle this setting and see all the animation is still animating. That feels bad. Right now we have to rely on developer or per widget to handle them seems not too scalable.

In the same time I wonder for the fade case, whether turning on disable animation should only turn off certain animation. Or maybe ticker may be used outside the just visual animation. Then in that case I agree hook up directly to ticker mode may not be a good idea. Maybe what we want is to hardcode all animation duration to zero if disableAnimations is on.

evantea commented:
@justinmc copied this commit into my local workspace, using only the "Remove Animations" system setting images, GIFs paused automatically (though only after rebuilding the view). Thanks for putting this together!

justinmc committed:
Image swapping without frame advancement

This is a combination of 5 commits, the older 4 being:
WIP Image swapping
WIP exploring swapping image sources and caching frames
WIP swapping works without frame advancement, but some tests are failing
WIP Works, but seems to be leaking

justinmc committed:
Trying to help pinpoint null check test failure message

justinmc committed:
Fix wasm web failure.

justinmc committed:
Well that didn't fix it. I'd guess the failing one is after SizedBox (dispose)

justinmc committed:
Merge branch 'master' into pause-gif

chunhtai reviewed (commented):
Just one comment on the cache management, the rest LGTM

justinmc commented:
@chunhtai I guess this only helps with Image, but that issue seems more broad. I will edit the description.

justinmc committed:
Merge branch 'master' into pause-gif

justinmc committed:
Revert "Image swapping without frame advancement"

This reverts commit 3665a79e64087e9a62f20816042ba30e1a0455eb.

justinmc committed:
Cacheless image swapping

justinmc committed:
Revert "[web] Fallback to CanvasKit when WebGL is not available (#173629)"

This reverts commit f63cb75520d25943e5824c552b9632255aa58c33.

I'm trying to get wasm tests to work locally again.

justinmc committed:
I simply was running this test on the web when it's not supported :(

justinmc committed:
Initial load tests

mdebbar reviewed (dismissed):
Is https://github.com/flutter/flutter/pull/176492/commits/ebb7ee964302c52931df379d4632097743bcc8c4 still needed or can it be reverted?

justinmc committed:
Reapply "[web] Fallback to CanvasKit when WebGL is not available (#173629)"

This reverts commit ebb7ee964302c52931df379d4632097743bcc8c4.

I reverted that as a test and forgot about it (it did succeed in letting
the tests pass, though). Let's see if this passes now with the latest
master, which I'll merge in after this commit.

justinmc committed:
Merge branch 'master' into pause-gif

mdebbar reviewed (commented):
Thanks @justinmc :)

justinmc committed:
Use _frameNumber instead of _hasReceivedFirstFrame.

justinmc committed:
Fix test detection.

A Google test was failing because it wasn't properly waiting for the
image to finish loading. It couldn't wait for the image because I was
removing the listener immediately after the first frame had come in,
before the test was able to start waiting for it. This commit fixes it
by waiting for the first 2 frames instead.

This is not ideal because gifs will advance one frame when swapping
their source. The real solution should probably be to fix the way that
Google tests wait for images to finish loading, but I was unable to
figure that out because I could not reproduce the problem locally.

I'll open an issue for this.

justinmc committed:
TODO

justinmc committed:
Analyzer

justinmc committed:
Merge branch 'master' into pause-gif

justinmc committed:
Merge branch 'master' into pause-gif

justinmc committed:
Merge branch 'master' into pause-gif

justinmc committed:
Revert "Fix test detection."

This reverts commit a28b26a1ac35f6eab2903f55aa026fec262b6921.

This should no longer be necessary if I can land a pending fix in
Google's internal tests.

justinmc commented:
I have a local commit to revert https://github.com/flutter/flutter/pull/176492/commits/a28b26a1ac35f6eab2903f55aa026fec262b6921 that I will push as soon as my fix for the Google tests is merged.

justinmc committed:
Merge branch 'master' into pause-gif

justinmc commented:
Google tests have been fixed in Google internal cl/829551684. Now triggering a re-run to confirm that they pass with this PR applied.

chunhtai commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -1124,10 +1148,12 @@ class _ImageState extends State<Image> with WidgetsBindingObserver {
-    } else {
+    _isPaused = !TickerMode.of(context) || (MediaQuery.maybeDisableAnimationsOf(context) ?? false);
+
+    if (_isPaused && _frameNumber != null) {
```
do you still need to keep track of number of frames? or can this be a boolean field?

justinmc replied:
I was about to change it but I realized it is used to give the frame number to frameBuilder: https://main-api.flutter.dev/flutter/widgets/ImageFrameBuilder.html

chunhtai reviewed (approved):
LGTM
